rfd/0232-desktop-access-protobuf-protocol.md

```diff
@@ -0,0 +1,596 @@
+represent boolean values anyhow.
+
+### Backward and Forward Compatibility
+By following protocol buffer best practices, we get forward and backward
+compatibility "for free" at least with respect to serialization/deserialization
+of individual messages. This is one of the main strengths of protocol buffers.
+
+### Familiarity
+Protocol buffers are used extensively at Teleport with existing automation for
+both Go and Typescript code generation.
+
+## Protocol Specification Updates
+Unless otherwise noted in this section, all existing TDP messages will simply
+be modeled as equivalent protocol buffer message representations. The behavior
+of the protocol will be largely the same aside from the updates below.
+See [all message definitions](#appendix-a-example-message-definitions) below.
+
+### Framing
+All TDPB messages will be preceded by a simple framing header consisting of
+two, unsigned, 32-bit, big-endian encoded integers representing the message
+type and message length respectively.
+
+```
+| message type (uint32) | message_length (uint32) | message_data []byte |
+```
+
+### Proxy/Agent Protocol Selection via ALPN
+The current TDP specification provides no handshake mechanism that we can use
+to upgrade from TDP to TDPB. Instead, we'll utilize ALPN to attempt to negotiate
+the use of TDPB. Modern Desktop agents will successfully negotiate
+`teleport-tdpb-1.0`, while older agents will fail negotiation entirely. This will
+signal the Teleport Proxy to use legacy TDP messages for the connection.
+
+ALPN may also be useful if we ever need to make breaking changes to the protocol,
+such as changes to the envelope message structure.
+
+### Web Client/Proxy Protocol Selection
+Typically, Teleport Agents and Clients are expected to connect to a Proxy instance whose
+version is greater than or equal to its own, however, the Desktop Access web client is a
+special case. During the rollout of a Proxy upgrade, "modern" Proxy instances will serve
+"modern" web clients who may in turn establish websocket connections to "legacy" Proxies.
+Likewise, "modern" Proxy instances may receive connections from "legacy" web clients.
+In order to gracefully handle this pathological upgrade scenario, the web client will need
+to support both TDP and TDPB implementations.
+
+By default, both the Proxy and web client will default to TDP for their leg of the session.
+A new "TDP upgrade" message will be added to the legacy TDP protocol, allowing the proxy
+to initiate an upgrade to TDPB. The Proxy will check for a new query parameter, `tdpb=<version>`,
+on the incoming HTTP upgrade request. If present, the Proxy knows that the client is capable
+of using TDPB and will issue the upgrade message. Any previously received messages are discarded,
+and the connection begins anew with TDPB. "Modern" web clients will supply this new
+query parameter and listen for the new upgrade, but continue with TDP as usual. 
```

Comment:
So the web client always speaks TDP unless the proxy it's talking to initiates the ugprade?

At which point in the flow is this upgrade message sent? I just want to understand what the web client's code changes will look like to accommodate this new message without causing any of the blocking issues we've had in the past.